### ISSUE TITLE
AttributeError: 'NoneType' object has no attribute 'primary_endpoint'

### ISSUE THREAD

**Issue by fedexist** [Azure/blobxfer#63]
Hi,

I'm trying to use the High Level API to upload files to Azure Blob and I started from [this](https://blobxfer.readthedocs.io/en/latest/80-blobxfer-python-library/) from the documentation. I think I'm missing something because I don't understand how come I'm getting this kind of error, which is thrown by `create_destination_id`

Here is the complete trace:

```
Traceback (most recent call last):
  File ".../upload.py", line 61, in <module>
    uploader.start()
  File "...\venv\lib\site-packages\blobxfer\operations\upload.py", line 1167, in start
    self._run()
  File "...\venv\lib\site-packages\blobxfer\operations\upload.py", line 1079, in _run
    create_destination_id(ase._client, ase.container, ase.name)
  File "...\venv\lib\site-packages\blobxfer\operations\upload.py", line 185, in create_destination_id
    return ';'.join((client.primary_endpoint, container, name))
AttributeError: 'NoneType' object has no attribute 'primary_endpoint'
```

These are the blobxfer parameters printed before the execution:

```
============================================
         Azure blobxfer parameters
============================================
         blobxfer version: 1.1.1
                 platform: Windows-10-10.0.16299-SP0
               components: CPython=3.6.4 azstor.blob=1.0.0 azstor.file=1.0.0 crypt=2.2.2 req=2.18.4
       transfer direction: local -> Azure
                  workers: disk=1 xfer=1 md5=1 crypto=0
                 log file: blobxfer.log
              resume file: None
                  timeout: connect=30.0 read=30.0
                     mode: block
                  skip on: fs_match=False lmt_ge=False md5=True
        delete extraneous: False
                overwrite: True
                recursive: True
            rename single: False
              access tier: hot
         chunk size bytes: 4194304
           one shot bytes: 33554432
         strip components: 0
         store properties: attr=True md5=True
           rsa public key: None
       local source paths: ..\sourcefolder
============================================

```

The code is like it follows, whereas config is a yaml file structured similarly to the sample configuration yaml [here](https://blobxfer.readthedocs.io/en/latest/sample_config.yaml).

```
    upload_options = config['upload'][0]['options']
    concurrency_dict = config['concurrency']
    skip_on_options = upload_options.pop('skip_on', None)
    store_file_properties = upload_options.pop('store_file_properties', None)
    endpoint = "core.windows.net"

    concurrency = blobxfer.api.ConcurrencyOptions(**concurrency_dict)
    timeout = blobxfer.api.TimeoutOptions(30.0, 30.0, 10)
    store_file_properties = blobxfer.api.FilePropertiesOptions(**store_file_properties)
    vectored_io = blobxfer.api.VectoredIoOptions(0, VectoredIoDistributionMode.Disabled)

    general_options = blobxfer.api.GeneralOptions(concurrency=concurrency, timeout=timeout, **config['options'])
    upload_options = blobxfer.api.UploadOptions(**upload_options,
                                                access_tier="hot",
                                                rsa_public_key=None,
                                                stdin_as_page_blob_size=None,
                                                vectored_io=vectored_io,
                                                store_file_properties=store_file_properties,
                                                strip_components=0)
    skip_on_options = blobxfer.api.SkipOnOptions(**skip_on_options)

    creds = blobxfer.api.AzureStorageCredentials(general_options)

    #for name, key in credentials_dict.items():
    creds.add_storage_account('storageaccountname',
                              'storageaccountkey',
                              endpoint)

    source = blobxfer.api.LocalSourcePath()
    source.add_path('../sourcefolder')

    destination = blobxfer.api.AzureDestinationPath()
    destination.add_path_with_storage_account('container', 'storageaccountname')

    spec = blobxfer.api.UploadSpecification(upload_options, skip_on_options, source)
    spec.add_azure_destination_path(destination)

    uploader = blobxfer.api.Uploader(general_options, creds, spec)
    uploader.start()

```

I feel like I'm missing something trivial. 
Any help is appreciated.

Thank you

**Comment by alfpark** [comment]
This call `destination.add_path_with_storage_account('container', 'storageaccountname')` must have a `creds` entry for `storageaccountname`. Please double check that is the case.

You can also test getting the associated storage account with:

```
creds.get_storage_account(destination[0].lookup_storage_account('container'))
```

**Comment by fedexist** [comment]
Hi @alfpark,

from what I see in the implementation of add_path_with_storage_account in `bloxfer/operations/azure/__init__.py`:

```
    def add_path_with_storage_account(self, remote_path, storage_account):
        # type: (SourcePath, str, str) -> None
        """Add a path with an associated storage account
        :param SourcePath self: this
        :param str remote_path: remote path
        :param str storage_account: storage account to associate with path
        """
        if len(self._path_map) >= 1:
            raise RuntimeError(
                'cannot add multiple remote paths to SourcePath objects')
        rpath = blobxfer.util.normalize_azure_path(remote_path)
        self.add_path(rpath)
        self._path_map[rpath] = storage_account
```

both the parameters should be strings? That's why I used the storage account string.

If I try to use `creds.get_storage_account(destination[0].lookup_storage_account('container'))` to check for the associated storage account, I get the following error:

```
Traceback (most recent call last):
  File ".../upload.py", line 55, in <module>
    print(creds.get_storage_account(destination.lookup_storage_account('container')))
  File "...\venv\lib\site-packages\blobxfer\operations\azure\__init__.py", line 84, in get_storage_account
    return self._storage_accounts[name]
KeyError: <blobxfer.operations.azure.StorageAccount object at 0x0000022CB4C85C88>
```

**Comment by alfpark** [comment]
That's correct, it should be a string type. 

Sorry, my check was off, try: `creds.get_storage_account(spec.destinations[0].lookup_storage_account('container'))`

**Comment by fedexist** [comment]
Ok, I checked and the output is correct: the lookup of the storage account allows me to access the credentials I've set (name, endpoint, key, clients and so on)

```
    sa = creds.get_storage_account(spec.destinations[0].lookup_storage_account('container'))
    print(sa.name, sa.endpoint, sa.key, sa.block_blob_client.primary_endpoint)
```
outputs correctly:

`storageaccountname core.windows.net storageaccountkey storageaccountname.blob.core.windows.net`

So, looking for the reason as to why ase._client would be None, I found the ase._client yielded by `_generate_destination_for_source` in `blobxfer/operations/upload.py` was None, because `populate_from_local` didn't assign the correct client to ase._client, as the mode passed "auto" (string) wasn't recognized as `StorageModes.Auto`.

So, in the end, it was a really silly oversight on my part and used `mode=StorageModes.Auto` when creating the UploadOptions.

Thank you, @alfpark, for your help.
